### DISCUSSION TITLE
How to add members / followers to quorum queue?

### DISCUSSION THREAD

**Discussion by peacock-el-gap** [rabbitmq/rabbitmq-server#8514] [Questions] [answered]
**Context**
RabbitMQ running on K8s, 3 nodes.
Our application after connecting to RabbitMQ cluster declares necessary quorum queues (no queue declaration using configuration file).
RabbitMQ version: 3.11.2

**Problem**
When RabbitMQ cluster starts, application connects to first node, not waiting for other nodes and declares quorum queue. As a result such queue is populated only on this first node. I'm looking for a way do add members / followers to such queue when they are ready and connected to cluser.
I know, that it is possible using `rabbitmq-queues` CLI.

**Questions**
1) Is it possible to add members / followers to quorum queue using API?
2) Is there a way to configure RabbitMQ cluster, that new cluster node will be automatically added to each auorum queue as a follower?
3) What is the best approach to ensure, that quorum queues are distributed on all nodes connected to cluster, even if such nodes are connected after queue declaration? Note, that I wan to avoid queue declaration by config file - I want my application to declare queue but I want to avoid situation, that my application impmenents functionality needed to observe cluster nodes readiness. I also want to avoid situation that tha application should wait till RabbitMQ cluster is ready and all nodes are ready.

**Comment by michaelklishin** [comment]
The default quorum queue target replica count is 3, which is perfectly sufficient for a lot of cases.
This replication factor [can be configured per queue](https://rabbitmq.com/quorum-queues.html#replication-factor).

There are also `rabbitmq-queues grow` and `rabbitmq-queues shrink`. There is also `rabbitmq-queues rebalance` which is supposed to be run after all cluster membership changes are complete, or all nodes are restarted during a rolling upgrade. See [Managing Replicas](https://rabbitmq.com/quorum-queues.html#replica-management).

**Reply by peacock-el-gap** [reply]
@michaelklishin,  thank you for your anwser 😃.

As I stated in my post I understand, that we can use `rabbitmq-queues` CLI.
It is also true, that even if the default quorum queue target replica count is 3, it is not applied if node count is less then 3 during queue declaration.

How about my questions?
Especially - can I use API to grow queue? Is it possible or not?
Unfortunatelly [API specification](https://rawcdn.githack.com/rabbitmq/rabbitmq-server/v3.12.0/deps/rabbitmq_management/priv/www/api/index.html) is not clear at this point 😢.

In [API specification](https://rawcdn.githack.com/rabbitmq/rabbitmq-server/v3.12.0/deps/rabbitmq_management/priv/www/api/index.html) we can find information, that we can use `PUT` method for `/api/queues/vhost/name endpoint`.

_To PUT a queue, you will need a body looking something like this:
{"auto_delete":false,"durable":true,"arguments":{},"node":"rabbit@smacmullen"}_

I've tried `PUT` with several options to add members / followers, but without success.

**Reply by mkuratczyk** [reply, marked as answer]
I don't think there is an API to achieve that.

There is a work-in-progress PR to periodically validate quorum queue membership:
https://github.com/rabbitmq/rabbitmq-server/pull/8218

The issue you are seeing is something we discussed in the past: https://github.com/rabbitmq/rabbitmq-server/issues/3850
I still believe we should not accept a declaration in the first place. The question remains what the behaviour should be (rejection, internal wait/retry that looks like a very slow declaration from the application's perspective or what)

**Reply by peacock-el-gap** [reply]
@mkuratczyk, thanx for your reply.
I agree that in such situation declaration should not be accepted. Rejection is my favourite behaviour, however I'm not sure, if I understand all consequences.

Topic closed.
